### PR TITLE
#1005: Fix XPI select field not populating correctly

### LINKED ISSUE #1005: Extension select element doesn't populate correctly
The 'extension' select element on the Opt Out Study fields does not populate correctly. Essentially, there is no Query component on the page to actually _load_ the extensions, which means that whatever is currently loaded in store is displayed in the select.

This can be reproduced by opening the `/recipe/new` page, switching to `opt-out-study`, and viewing the XPI select field. The field will be empty, until you navigate to `/extension` and actually load the list.

This should be resolved by adding a Query component near the extension select element on the Opt Out fields.

### PR DESCRIPTION
**Pull request by andymikulski**
Fixes #1005 

Basically, the issue here is that there was no query component on the recipe form. Any values displayed in the select field were taken from whatever was already loaded in the store. This PR adds a `PaginatedSelect` base component which handles loading paginated data and inserting it into a `Select` field.

- Adds `PaginatedSelect` abstract component
  - Given a list and query component, displays a `show more...` button in the list until all pages are loaded and displayed in the select
- Adds `ExtensionSelect`, `RecipeSelect` components. Technically `RecipeSelect` is unnecessary, but I wanted to make sure the base component would work for both use cases.
- Adds `numPages` and `all` properties to Recipes and Extension `listing` stores
- Adds appropriate selectors

screen shot 2017-08-24 at 10 21 36 am

This PR requires tests, but the rest is up for feedback!

### CODE REVIEW

File: recipe-server/client/control_new/state/app/recipes/reducers.js [hunk 1]
```diff
@@ -67,11 +68,23 @@ function items(state = new Map(), action) {
...
   switch (action.type) {
-    case RECIPE_PAGE_RECEIVE:
+    case RECIPE_PAGE_RECEIVE: {
+      let totalRecipes = state.get('all', new Set());
+
+      fromJS(action.recipes.results.forEach(result => {
+        totalRecipes = totalRecipes.add(fromJS(result));
+      }));
+
+      totalRecipes = totalRecipes.sort((a, b) => a.get('id') <= b.get('id'));
```
Comment (mythmon) [inline recipe-server/client/control_new/state/app/recipes/reducers.js]: This isn't correct. The comparator function passed to `.sort()` [should return -1, 0, or 1](https://facebook.github.io/immutable-js/docs/#/Set/sort), not a boolean. Assuming `id` is a number, `a.get('id') - b.get('id')` would would.

File: recipe-server/client/control_new/state/app/extensions/reducers.js [hunk 2]
```diff
@@ -24,11 +24,22 @@ function items(state = new Map(), action) {
 
 function listing(state = new Map(), action) {
   switch (action.type) {
-    case EXTENSION_PAGE_RECEIVE:
+    case EXTENSION_PAGE_RECEIVE: {
+      let totalExtensions = state.get('all', new Set());
+
+      fromJS(action.extensions.results.forEach(result => {
```
Comment (mythmon) [inline recipe-server/client/control_new/state/app/extensions/reducers.js]: What does `fromJS` do here?

File: recipe-server/client/control_new/components/recipes/OptOutStudyFields.js [hunk 3]
```diff
@@ -61,13 +53,15 @@ export default class OptOutStudyFields extends React.Component {
...
-              </Select>
+              <ExtensionSelect disabled={disabled} />
+            </FormItem>
+
+            <FormItem
+              label="Add-on Extension"
+              name="arguments.addonUrl"
+              initialValue={recipeArguments.get('addonUrl', '')}
+            >
+              <RecipeSelect disabled={disabled} />
```
Comment (mythmon) [inline recipe-server/client/control_new/components/recipes/OptOutStudyFields.js]: While useful for testing, this probably shouldn't be here in the final version.

**Comment by andymikulski** [comment]
>As an alternative approach, would it make sense to load all extensions in this list?

This is the most desirable path, though I was just working with the APIs that are available. If @marcrowo wants to hop on this and create an endpoint that lists all extensions, that'd remove the need for 98% of this PR.

**Comment by mythmon** [comment]
I don't think we should make an endpoint that lists all extensions. Or rather, if we do, we shouldn't *also* have the paginated API. We should pick one. The action creator that fetches extensions can make multiple queries to the API, and keep requesting pages until it has everything.

**Comment by andymikulski** [comment]
Based on yesterday's IRC conversation, it sounds like there are two API solves for this:

1. Change the existing paginated endpoint to allow returning the entire list in one call
1. Add filters to the 'extension listing' endpoint to allow text search (like on the Recipes Listing)

If @marcrowo can handle one of those, this can be wrapped up. I'm a fan of the `filters` approach since it adds more features to other parts of the site for the same amount of work.

**Comment by rehandalal** [comment]
+1 for filters

**Comment by andymikulski** [comment]
Taking this over now that @marcrowo 's filter stuff has landed!

**Comment by andymikulski** [comment]
@mythmon I've added `filter` support for fetching extensions and added the `ExtensionSelect` component, which shows the first page of extensions, unless a text search is provided.

This is ready for review!

File: recipe-server/client/control_new/components/extensions/ExtensionSelect.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        search,
+      });
+    }, 200);
+  }
+
+  render() {
+    const { search } = this.state;
+    let displayedList = this.props.extensions;
+
+
```
Comment (mythmon) [inline recipe-server/client/control_new/components/extensions/ExtensionSelect.js]: Nit: extra line here.

File: recipe-server/client/control_new/components/extensions/ExtensionSelect.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    if (this.props.isLoadingSearch) {
+      displayedList = new List();
+    }
+
+    return (
+      <div>
+        <QueryMultipleExtensions filters={queryFilters} pageNumber={1} />
+        <Select
+          filterOption={false}
+          placeholder="Select Extension"
```
Comment (mythmon) [inline recipe-server/client/control_new/components/extensions/ExtensionSelect.js]: Overall, It isn't obvious that this i s a search field. Can we change this to something like "Search for Extension", to make it clear that users should type? Maybe a search icon would be helpful?

File: recipe-server/client/control_new/less/main.less [hunk 6]
```diff
@@ -188,3 +188,7 @@ fieldset {
     max-width: 300px;
   }
 }
+
+.select-show-more {
+  font-weight: 900;
+}
```
Comment (mythmon) [inline recipe-server/client/control_new/less/main.less]: I don't think this is used anymore?

File: recipe-server/client/control_new/state/app/extensions/reducers.js [hunk 7]
```diff
@@ -24,11 +24,22 @@ function items(state = new Map(), action) {
...
   switch (action.type) {
-    case EXTENSION_PAGE_RECEIVE:
+    case EXTENSION_PAGE_RECEIVE: {
+      let totalExtensions = state.get('all', new Set());
+
+      action.extensions.results.forEach(result => {
+        totalExtensions = totalExtensions.add(fromJS(result));
+      });
+
+      totalExtensions = totalExtensions.sort((a, b) => a.get('id') <= b.get('id'));
```
Comment (mythmon) [inline recipe-server/client/control_new/state/app/extensions/reducers.js]: This sort method won't work. Immutable's `.sort()` method (and JS's as well), want a comparator that returns `0`, `1`, or `-1`.

File: recipe-server/client/control_new/state/app/extensions/selectors.js [hunk 8]
```diff
@@ -7,11 +7,13 @@ export function getExtension(state, id, defaultsTo = null) {
   return state.app.extensions.items.get(id, defaultsTo);
 }
 
-
 export function getExtensionListingCount(state) {
   return state.app.extensions.listing.get('count');
 }
 
+export function getExtensionPageCount(state, defaultsTo = 0) {
```
Comment (mythmon) [inline recipe-server/client/control_new/state/app/extensions/selectors.js]: Are we still using this? Do you want to keep it anyways?

File: recipe-server/client/control_new/state/app/extensions/reducers.js [hunk 9]
```diff
@@ -24,11 +24,22 @@ function items(state = new Map(), action) {
...
+
       return state
         .set('count', action.extensions.count)
+        .set('numPages', Math.ceil(action.extensions.count / 10))
         .set('pageNumber', action.pageNumber)
-        .set('results', fromJS(action.extensions.results.map(extension => extension.id)));
+        .set('results', fromJS(action.extensions.results.map(extension => extension.id)))
+        .set('all', totalExtensions);
```
Comment (mythmon) [inline recipe-server/client/control_new/state/app/extensions/reducers.js]: Do we still use `all` or `totalExtensions` anywhere? I don't see it.

File: recipe-server/client/control_new/state/app/recipes/selectors.js [hunk 10]
```diff
@@ -42,6 +46,10 @@ export function getRecipeListingPageNumber(state) {
   return state.app.recipes.listing.get('pageNumber');
 }
 
+export function getRecipePageCount(state, defaultsTo = 0) {
```
Comment (mythmon) [inline recipe-server/client/control_new/state/app/recipes/selectors.js]: Do you want to keep these recipes changes?

File: recipe-server/client/control_new/components/extensions/ExtensionSelect.js (hunk 5 above)
Comment (andymikulski) [reply recipe-server/client/control_new/components/extensions/ExtensionSelect.js]: How do you feel about this treatment:

screen shot 2017-09-05 at 9 29 06 am

**Comment by andymikulski** [comment]
I've taken out the unnecessary code from earlier changes and corrected feedback items. This should be ready for a final pass!

File: recipe-server/client/control_new/components/extensions/ExtensionSelect.js (hunk 5 above)
Comment (mythmon) [reply recipe-server/client/control_new/components/extensions/ExtensionSelect.js]: 👍

**Comment by mythmon** [comment]
TC failure is #1003, which has been fixed on master. Merging without it.

**Pull request merged**